float32 step
float32 rev
string direction